### ISSUE TITLE
The Question about the mask of window attention

### ISSUE THREAD

**Issue by MARD1NO** [microsoft/Swin-Transformer#38]
Nice work!And i reading your code recently. But i cannot understand well about the implementation of the mask in shifted window attention. 

I simply draw a picture like below. The red mean the mask, and i choose windowsize as 2, shiftsize as 1. 

I think the mask should be like this
![image](https://user-images.githubusercontent.com/42901638/115377048-d361c880-a201-11eb-8be6-e31d007a0bf9.png)
but i use your code to generate mask like this: 

```python
import torch
import torch.nn as nn


def window_partition(x, window_size):
    """
    Args:
        x: (B, H, W, C)
        window_size (int): window size

    Returns:
        windows: (num_windows*B, window_size, window_size, C)
    """
    B, H, W, C = x.shape
    x = x.view(B, H // window_size, window_size, W // window_size, window_size, C)
    windows = x.permute(0, 1, 3, 2, 4, 5).contiguous().view(-1, window_size, window_size, C)
    return windows


window_size = 2
shift_size = 1
H, W = 4, 4
img_mask = torch.zeros((1, H, W, 1))  # 1 H W 1
h_slices = (slice(0, -window_size),
            slice(-window_size, -shift_size),
            slice(-shift_size, None))
w_slices = (slice(0, -window_size),
            slice(-window_size, -shift_size),
            slice(-shift_size, None))

cnt = 0
for h in h_slices:
for w in w_slices:
img_mask[:, h, w, :] = cnt
cnt += 1

mask_windows = window_partition(img_mask, window_size)  # nW, window_size, window_size, 1
mask_windows = mask_windows.view(-1, window_size * window_size)
attn_mask = mask_windows.unsqueeze(1) - mask_windows.unsqueeze(2)
attn_mask = attn_mask.masked_fill(attn_mask != 0, float(-100.0)).masked_fill(attn_mask == 0, float(0.0))
attn_mask = attn_mask.unsqueeze(1).unsqueeze(0)

"""
tensor([[[[[   0.,    0.,    0.,    0.],
           [   0.,    0.,    0.,    0.],
           [   0.,    0.,    0.,    0.],
           [   0.,    0.,    0.,    0.]]],


         [[[   0., -100.,    0., -100.],
           [-100.,    0., -100.,    0.],
           [   0., -100.,    0., -100.],
           [-100.,    0., -100.,    0.]]],


         [[[   0.,    0., -100., -100.],
           [   0.,    0., -100., -100.],
           [-100., -100.,    0.,    0.],
           [-100., -100.,    0.,    0.]]],


         [[[   0., -100., -100., -100.],
           [-100.,    0., -100., -100.],
           [-100., -100.,    0., -100.],
           [-100., -100., -100.,    0.]]]]])
"""
```
I cannot understand it, can you give me a favor?

**Comment by MARD1NO** [comment]
emmm I think i'm wrong. I draw a new pic for the mask attention
![image](https://user-images.githubusercontent.com/42901638/115496492-5d5e6f80-a29c-11eb-8cbf-76aff3e40d7e.png)

**Comment by zeliu98** [comment]
Hi, it seems all right. I simply visualize the `img_mask` and `attn_mask` using image-size=14x14, window-size=7x7, shift=3.
![image](https://user-images.githubusercontent.com/46211709/115505327-d7e2bb80-a2ab-11eb-8e90-ec27b8817c28.png)
Below is the visualization code:
```python
import torch

import matplotlib.pyplot as plt


def window_partition(x, window_size):
    """
    Args:
        x: (B, H, W, C)
        window_size (int): window size

    Returns:
        windows: (num_windows*B, window_size, window_size, C)
    """
    B, H, W, C = x.shape
    x = x.view(B, H // window_size, window_size, W // window_size, window_size, C)
    windows = x.permute(0, 1, 3, 2, 4, 5).contiguous().view(-1, window_size, window_size, C)
    return windows


window_size = 7
shift_size = 3
H, W = 14, 14
img_mask = torch.zeros((1, H, W, 1))  # 1 H W 1
h_slices = (slice(0, -window_size),
            slice(-window_size, -shift_size),
            slice(-shift_size, None))
w_slices = (slice(0, -window_size),
            slice(-window_size, -shift_size),
            slice(-shift_size, None))
cnt = 0
for h in h_slices:
    for w in w_slices:
        img_mask[:, h, w, :] = cnt
        cnt += 1

mask_windows = window_partition(img_mask, window_size)  # nW, window_size, window_size, 1
mask_windows = mask_windows.view(-1, window_size * window_size)

attn_mask = mask_windows.unsqueeze(1) - mask_windows.unsqueeze(2)
attn_mask = attn_mask.masked_fill(attn_mask != 0, float(-100.0)).masked_fill(attn_mask == 0, float(0.0))

plt.matshow(img_mask[0, :, :, 0].numpy())
plt.matshow(attn_mask[0].numpy())
plt.matshow(attn_mask[1].numpy())
plt.matshow(attn_mask[2].numpy())
plt.matshow(attn_mask[3].numpy())

plt.show()

```

**Comment by MARD1NO** [comment]
> Hi, it seems all right. I simply visualize the `img_mask` and `attn_mask` using image-size=14x14, window-size=7x7, shift=3.
> ![image](https://user-images.githubusercontent.com/46211709/115505327-d7e2bb80-a2ab-11eb-8e90-ec27b8817c28.png)
> Below is the visualization code:
> 
> ```python
> import torch
> 
> import matplotlib.pyplot as plt
> 
> 
> def window_partition(x, window_size):
>     """
>     Args:
>         x: (B, H, W, C)
>         window_size (int): window size
> 
>     Returns:
>         windows: (num_windows*B, window_size, window_size, C)
>     """
>     B, H, W, C = x.shape
>     x = x.view(B, H // window_size, window_size, W // window_size, window_size, C)
>     windows = x.permute(0, 1, 3, 2, 4, 5).contiguous().view(-1, window_size, window_size, C)
>     return windows
> 
> 
> window_size = 7
> shift_size = 3
> H, W = 14, 14
> img_mask = torch.zeros((1, H, W, 1))  # 1 H W 1
> h_slices = (slice(0, -window_size),
>             slice(-window_size, -shift_size),
>             slice(-shift_size, None))
> w_slices = (slice(0, -window_size),
>             slice(-window_size, -shift_size),
>             slice(-shift_size, None))
> cnt = 0
> for h in h_slices:
>     for w in w_slices:
>         img_mask[:, h, w, :] = cnt
>         cnt += 1
> 
> mask_windows = window_partition(img_mask, window_size)  # nW, window_size, window_size, 1
> mask_windows = mask_windows.view(-1, window_size * window_size)
> 
> attn_mask = mask_windows.unsqueeze(1) - mask_windows.unsqueeze(2)
> attn_mask = attn_mask.masked_fill(attn_mask != 0, float(-100.0)).masked_fill(attn_mask == 0, float(0.0))
> 
> plt.matshow(img_mask[0, :, :, 0].numpy())
> plt.matshow(attn_mask[0].numpy())
> plt.matshow(attn_mask[1].numpy())
> plt.matshow(attn_mask[2].numpy())
> plt.matshow(attn_mask[3].numpy())
> 
> plt.show()
> ```

Thank so much!!!!!!!!!!!!! It is such a great idea to improve the efficiency of Shift window Attention!!!!!!!!!!!

**Comment by abhi-glitchhg** [comment]
Thanks @MARD1NO ; I loved your blog too! It helped me to understand!

**Comment by zeliu98** [comment]
Hi @HC-2016, thanks for pointing it. A refined figure is here:
![image](https://user-images.githubusercontent.com/46211709/145946588-7d05045a-5964-43f2-b82a-0128537fa6f4.png)

**Comment by YangGangZhiQi** [comment]
Hi, Great Work! I have a question about using attn_mask. 
![attn_mask疑问](https://user-images.githubusercontent.com/87355944/146151808-54860dec-b03b-4ef2-918b-4083dbb9ce35.png)

In Line133, why do we just add attn_mask to attn, not multiply?  Using add operation let me think the atnn_mask is just a bias like relative_position_bias in Line 129.
I am a beginner in Swin Transformer.  I would appreciate it if  someone could help me.

**Comment by piantou** [comment]
> Hi, Great Work! I have a question about using attn_mask. ![attn_mask疑问](https://user-images.githubusercontent.com/87355944/146151808-54860dec-b03b-4ef2-918b-4083dbb9ce35.png)
> 
> In Line133, why do we just add attn_mask to attn, not multiply? Using add operation let me think the atnn_mask is just a bias like relative_position_bias in Line 129. I am a beginner in Swin Transformer. I would appreciate it if someone could help me.

Just like [0.5 ,0.7, 0.3]+[0, -100, 0]=[0.5, -99.3, 0.3] , will ignore the second value 0.7

**Comment by fedral** [comment]
Here is how I understand of masked attention:
 
Due to the circle shift operation of input feature,  areas of diffenent semantic  are  combined together into one window.
As shown in window 3,  area 4, 5, 7, 8 in the same window;  

 During the calculation of  (N,N) dot product matrix , the mask confines weights of area 4 only calculated with area 4 positions, others spatial positions(5, 7, 8 ) are set to small vaule near 0 through the trick of softmax(-100.0).

In sum, for windows with differenent and multiple sematic areas, the mask confines window attention within each area independently.  ^-^

**Comment by myexceptions** [comment]
asdfadsf
